front page: Who is Nix for? (#872)
NixOS/nix.dev: source/index.md

- Atomic upgrades and rollbacks.

## Who is Nix for?

> Nix is a tool for software developers.

Reviewer: ```suggestion
Nix is primarily a tool for software developers.
```

I think adding "primarily" helps keep the door open to different audiences who may be willing enough to learn and use Nix, while not distracting from the focus on software developers.

> Nix is primarily a tool for software developers.

Reviewer 2: Shouldn't this be a PR to nixos.org instead? nix.dev is for learning about Nix, but this section doesn't teach anything, it's just marketing. If somebody went onto nix.dev, we should assume that they want to learn about Nix, so I don't think this is necessary or useful here.

And if your intention is to make Nix platform _contributors_ be more focused on improving the developer story, I also don't think this would achieve that. Nix is a community of various individuals and companies, which will do whatever they're most interested in (which often isn't focused on the developer story).

Author: > Shouldn't this be a PR to nixos.org instead?

In an ideal world, maybe? In practice, I find that code unmaintainable and refuse to work with it.

> nix.dev is for learning about Nix, but this section doesn't teach anything, it's just marketing.

This is very clearly not marketing, because it doesn't advertise anything, it doesn't try to convince anyone of anything. The purpose is, as said in the commit message, to set expectations and prepare people for what they'll deal with. The phrase "going back to a world without Nix" is only a measure to avoid sounding academic. The technically correct statement would be "According to community surveys 2022 and 2023, people who identify with these descriptions very probably find Nix useful".

> And if your intention is to make Nix platform contributors be more focused on improving the developer story, I also don't think this would achieve that.

While that's not my intention, I strongly doubt it's true what you say. [The first impression matters a lot](https://en.wikipedia.org/wiki/Priming_(psychology)), and even if this particular piece is below the fold, whoever cares enough to read it will carry that message around with them for a long time to shape their thinking.

Reviewer 2: > This is very clearly not marketing

Fair yeah, but still, setting expectations seems like a main landing-page kind of deal. And you refusing to touch nixos.org is clearly something gone wrong that we should prioritise addressing instead. This is definitely worth a @NixOS/marketing-team ping.

Reviewer 2: I guess it shouldn't hurt to _also_ set expectations for nix.dev specifically, but so far this PR seems to set expectations for Nix in general.

Author: I'm being deliberately generic here, because we don't have that sort of writing anywhere. I'd rather have it here now and move it around later. I've been repeating myself and trying to come up with things that help orient people for almost two years now, getting bogged down with concerns on every step. I'm all in favor of being considerate with far-reaching decisions, but *come on!*

Reviewer: I'm all in favor of getting more content written as well! For now, I'd rather have redundant information in different places than try to keep it in a single place (as long as that information won't go out of date soon). It's been my experience that people find nix through multiple different links, and often don't visit all the other links (e.g. they find combinations of the nix pills, manuals, nix.dev, wiki, and multiple other sites, but often don't find all of those), so having repeating information (which doesn't go out of date soon) is better IMO at the moment.

Properly organising all the content under a single structure takes way more time (but we're moving there, e.g. with the nix.dev -> docs.nixos.org rename and so on), and when we get there we can start cutting the redundant content.

My only concern is making sure what we're adding is:
1. Not confusing for the intended audience
2. Improving on the current content

I think the latest version of this PR fits both criteria, so I'm approving!

> You don't have to be a professional and you don't need formal education in informatics to greatly benefit from Nix.

Reviewer: This sentence should come after the next one: first clarify what we mean by software developer, then make this statement that you don't have to be a professional and so on...

Author: Yeah, I was shuffling around those sentences until they stopped making sense to me. You're probably right.

> For our purposes here, you are a software developer if you manipulate character strings to make computers do exactly what you want.

Reviewer: I think this sentence doesn't help narrow down the audience.

I'm coming from this with mostly an assumption (roughly reinforced by answers in the Nix community surveys): that there are primarily two types of audience that come to Nix/NixOS. The first audience is people who actually write code and build software (I'm not using the term "software developer" here to avoid confusion with the current version of the PR). The second audience is people who are Linux enthusiasts - they use a Linux OS, know how to perform certain tasks on the terminal and are usually a bit more tech-savvy, but they are "script kiddies" at best when it comes to actual understanding about how software works.

With this assumption in mind, "manipulate character strings to make computers do exactly what you want" can be applied to both types of audience, but as I interpreted it from the other sources you listed, the focus is only on the first type of audience, not the second.

```suggestion
For our purposes here, you are a software developer if you write or build code for computers.
```

I feel this suggestion is excluding some part of the audience that is intended to be included, but I am not sure about that. I made this suggestion to start from the other end: very narrow, and hoping that we can identify other audiences that we want to include and keep adjusting the sentence.

Author: Fully agreed that there's a producer <-> consumer spectrum, but I really didn't want to go down that road here. 

My way of phrasing it is certainly peculiar. It's more to accommodate a sentiment expressed in https://github.com/NixOS/nixos-homepage/issues/828: I want the community to be welcoming to nerds of all sorts, and especially beginners. If you had asked me 10 years ago, I would not have identified with your definition. Beginners don't *write* code. They use it, they configure existing software (manipulating what isn't or what we usually don't consider code), sometimes read code, and then maybe start changing that code. The cool thing about Nix is that, while certainly you'll end up writing code, in the beginning it doesn't really feel like it.

The other thing I was trying to capture is that what you really need is a particular mind set (and a lot of time to learn things) rather than a particular occupation (which "writing code" is) or qualification. We eventually meet at what ends up being programming. It's just that at the boundaries there are enough people who don't actually want to program - but still do because they have to deal with the fact that software sucks or doesn't do what they want. Natural scientists are a good example. All the anecdotes I know  amount to "we write code [that breaks] all the time but would rather do `$SCIENCE`".

All this clearly excludes people who won't use the shell or a plain text editor.

That was my take at "not too wide, not too narrow".

Reviewer 2: > All this clearly excludes people who won't use a command line interface or a plain text editor.

This sounds much more straightforward, how about just mentioning this instead?

Author: ```suggestion
For our purposes here, you are a software developer if you use a command line interface and a plain text editor.
```

You mean like this?

Reviewer: I see what you were going for now. From the PR description, I thought you wanted to closely align with the marketing team's audience focus (which seems to be closer to the definition I suggested).

If the idea is to keep it broader, why not remove the term "software developer" completely? It's bound to cause confusion (probably similar to mine).

My suggestion would be to shorten this to something like: "Nix is a tool for people comfortable with a command line interface and a plain text editor."

Author: > Nix is a tool for people comfortable with a command line interface and a plain text editor.

Hm, I agree it's more straightforward and descriptive. But also completely lacks the part that I find most important, and that's about the spirit of it all, the "why does it matter". Why would you bother going through pages of dry text for hours on end? Your suggestion makes it sound like this is yet another command line tool. I'm getting flashbacks from the vision statement discussion...

Okay, which group of people would care about the following?

- Makes computers *actually* do (well, almost) exactly the same thing every time (it's cached `execve()` in a sandbox on immutable files)
- Allows expressing process execution and file system operations in terms of declarative programs
- Exposes and allows managing all the complexity of the open source software ecosystem

If not forced by necessity, probably it's perfectionist control-freaks. Maybe that's the uncomfortable truth that prevents us from saying it out loud. If there's a constructive variant of it, I'd still take that.
